### PR TITLE
o/devicestate/devicemgr: register install-device hook, run if present in install

### PR DESCRIPTION
**Pull request by anonymouse64**
Run a new install-device hook during install mode after setup-run-system, but
before the ensure-next-boot-to-run-mode task.

Also fix a subtle bug in the addTask() function, it wasn't setting the previous
task properly, so all new tasks would just depend on the original
setup-run-system task, and not on the next task in the chain.

I can add a spread test here executing an install-device hook, but absent putting the gadget snap in devmode so that the install-device hook can touch things, there's not really anything the hook could do that we could measure since the state.json for install mode disappears and anywhere the snap could currently write to is also ephemeral.

### CODE REVIEW

File: overlord/devicestate/devicestate_serial_test.go [hunk 1]
```diff
@@ -1882,10 +1882,16 @@ func (s *deviceMgrSerialSuite) TestDeviceRegistrationNotInInstallMode(c *C) {
 	st.Set("seeded", true)
 	// set run mode to "install"
 	devicestate.SetSystemMode(s.mgr, "install")
+
+	// one of the startup managers needs to access the gadget snap to proceed
+	r2 := devicestatetest.MockGadget(c, s.state, "pc", snap.R(2), nil)
```
Comment (pedronis) [inline overlord/devicestate/devicestate_serial_test.go:1887]: shouldn't we set ensureInstalledRan instead perhaps? or did this test always run install itself?

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -931,6 +948,22 @@ func (m *DeviceManager) ensureInstalled() error {
...
+		summary := i18n.G("Run install-device hook")
+		hooksup := &hookstate.HookSetup{
+			// TODO: what's a reasonable timeout for the install-device hook?
+			Snap: model.Gadget(),
+			Hook: "install-device",
+		}
+		installDevice := hookstate.HookTask(m.state, summary, hooksup, nil)
+		addTask(installDevice)
+		// hooks are under a different manager, make sure we consider
+		// it immediately
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:963]: this is not true anymore, there is only one taskrunner configured after all the other managers, so likely don't need the EnsureBefore anymore, probably same for a couple of others in this file

**Review by pedronis — COMMENTED** [review]
last commits look good, couple remarks

File: overlord/devicestate/devicestate_serial_test.go (hunk 1 above)
Comment (anonymouse64) [reply overlord/devicestate/devicestate_serial_test.go:1887]: this test was always running install itself, but mocking ensureInstalledRan seems more logical

File: overlord/devicestate/devicemgr.go (hunk 2 above)
Comment (anonymouse64) [reply overlord/devicestate/devicemgr.go:963]: thanks for pointing this out, I will clean this up in the other instances too

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -939,6 +945,19 @@ func (m *DeviceManager) ensureInstalled() error {
 	}
 	prev = setupRunSystem
 
+	// add the install-device hook before ensure-next-boot-to-run-mode if it
+	// exists in the snap
+	if hasInstallDeviceHook {
+		summary := i18n.G("Run install-device hook")
+		hooksup := &hookstate.HookSetup{
+			// TODO: what's a reasonable timeout for the install-device hook?
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:953]: the default is 10 mins nowadays (I thought it was shorter but was wrong), which seems fair not having any example yet

File: overlord/devicestate/devicemgr.go [hunk 4]
```diff
@@ -169,6 +170,16 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	return m, nil
 }
 
+type genericHook struct{}
```
Comment (mvo5) [inline overlord/devicestate/devicemgr.go:173]: Nice!

**Review by mvo5 — APPROVED** [review]
This looks great, thanks you!

But I'm slightly confused about the PR description. It says: `Also fix a subtle bug in the addTask() function,...` but I don't see this in the PR? Am I overlooking something?

**Comment by anonymouse64** [comment]
> It says: Also fix a subtle bug in the addTask() function,... but I don't see this in the PR? Am I overlooking something?

Sorry I ended up picking that fix up in the previous PR: https://github.com/snapcore/snapd/pull/10035/commits/73e4cb21bf7c7d0f54ba1205beb9ad866c017845#diff-3dedbc3a1986cf905aca293ca6d16560d297253859ecaa8bcea2b7dab7143b3cR938

**Pull request merged**